rfd/0152-automatic-database-users-mongodb.md

```diff
@@ -0,0 +1,202 @@
+Differences will be outlined in the sections below.
+
+### The admin user connection
+
+The Database Service will connect as an admin user in order to manage database
+users. The admin user requires the following privileges:
+```json
+{
+  "createUser": "CN=teleport-admin",
+  "roles": [
+    { "role": "clusterMonitor", "db": "admin" },
+    { "role": "userAdminAnyDatabase", "db": "admin" }
+  ]
+}
+```
+
+As the implementation of other databases, the name of admin user is
+defined in the database spec `admin_user.name`, and the admin user will
+authenticate using X.509 for self-hosted databases.
+
+However, there is NO concept of "Stored Procedures" in MongoDB. The user
+provisioning logic will be carried through multiple `runCommand` calls
+implemented in Go. Multiple parallel database sessions will not race thanks to
+[semaphore
+locking](https://github.com/gravitational/teleport/blob/master/rfd/0113-automatic-database-users.md#locking).
+
+### Roles
+
+Unlike MySQL or PostgreSQL where roles are scoped to the entire database
+instance/cluster, a role in MongoDB is scoped to a specific database.
+
+For example, a custom role `myCustomRole` can be created on database `db1` with
+specific privileges on `db1`, and another role `myCustomRole` can be created on
+database `db2` with specific privileges on `db2`. They will be considered two
+different roles.
+
+Most built-in roles are available on all databases, while some built-in roles
+like `readAnyDatabase` can only be applied on the `admin` database.
+
+Therefore, when specifying database roles to assign for the user, it must be in
+the format of `<role-name>@<db-name>` to fully identify the role:
+
+```yaml
+kind: "role"
+version: "v6"
+metadata:
+  name: "example"
+spec:
+  options:
+    create_db_user_mode: keep
+  allow:
+    db_names:
+    - "db1"
+    - "db2"
+    - "db3"
+    db_roles: 
+    - "readAnyDatabase@admin"
+    - "readWrite@db2"
+    - "myCustomRole@db3"
+```
+
+Note that in the above example, even though the user is assigned role
+`readAnyDatabase@admin`, Teleport will block access to databases not in the
+`db_names`.
+
+### User accounts
+
+New users with name `CN=<teleport-username>` will be created on `$external`
+database to use X.509 authentication.
+
+All auto-provisioned users will be assigned role `teleport-auto-user@admin`.
+The `teleport-auto-user` is created on database `admin` with empty privileges.
+
+There is no built-in way to lock an user account in MongoDB, for deactivation
+purpose. As a workaround, the following `authenticationRestrictions` will be
+applied to the user account, in addition to stripping the roles:
+
+```json
+{
+  "updateUser": "CN=<teleport-username>",
+  "roles": [
+    { "role": "teleport-auto-user", "db": "admin" }
+  ],
+  "authenticationRestrictions": [
+    { "clientSource": ["0.0.0.0"] }
+  ]
+}
+```
+
+Limiting the `clientSource` effectively locks out the user from logging in.
```

Comment:
For the `clientSource` to take effect, does the MongoDB server need to be configured in any way? Or does it always take effect by default? Let's mention this here.

Reply:
Maybe I missed it in the RFD, but it doesn't look like we still mention anywhere whether client source restrictions always apply or have to be turned on via some MongoDB server configuration.

Reply:
yes. it's built-in and always checked. Updated now:
```
However, MongoDB has builtin
[`authenticationRestrictions`](https://www.mongodb.com/docs/manual/reference/method/db.createUser/#authentication-restrictions)
that restricts logins by `clientSource` or `serverAddress`, which is always
checked when a database user is being authenticated.
```